Documentation.app/Contents/MacOS/FP concepts.playground/Pages/Rank-N polymorphism.xcplaygroundpage/Contents.swift

```diff
@@ -0,0 +1,155 @@
+ */
+print(anotherRandomWordGenerator())
+/*:
+ And we can obviously pass a value of `RandomWordGenerator` around, which will always hold the `length` variable it needs. So we see that our struct behaves very much like a closure. Indeed, when the Swift compiler finds a closure definition in your code, it creates something similar to this struct under the hood.
+
+ ## Generic closures
+
+ Recall the `singletonArray` function we defined earlier:
+ ```
+ func singletonArray<T>(_ v: T) -> [T] {
+     [v]
+ }
+ ```
+ Can we write this function as a struct, like we explained in the previous section? Sure thing!
+ ```
+ struct SingletonArrayFunction {
+     func callAsFunction<T>(_ v: T) -> [T] {
+         [v]
+     }
+ }
+ ```
+ */
+/*:
+ We just expressed a polymorphic function as a struct!
+
+ Recall the function `callTwice` we defined earlier with our made up syntax:
+ ```
+ func callTwice<A, B>(_ f: <T>(T) -> [T], _ a: A, _ b: B) -> ([A], [B]) {
+     (f(a), f(b))
+ }
+ ```
+ Now that we have expressed a polymorphic function as a struct, can we express this function with valid Swift syntax? Maybe something like this?
+ ```
+ func callTwice<A, B>(_ f: SingletonArrayFunction, _ a: A, _ b: B) -> ([A], [B]) {
+     (f(a), f(b))
+ }
+ ```
+ */
+/*:
+ Well, now we can only pass a value of `SingletonArrayFunction`, which is to say that we can only pass one specific function. This is not what we wanted, we wanted to be able to pass any polymorphic function with this signature `<T>(T) -> [T]`. We must find another way.
+
+ ## Introducing FunctionK
+
+ `FunctionK` is a class found in Bow that represents any polymorphic function. `FunctionK` takes two type parameters that represent type constructors. Specifically, `FunctionK` represents functions from `Kind<F, T>` to `Kind<G, T>` that is polymorphic on `T`, or expressed with the made up notation we introduced earlier: `<T>Kind<F, T> -> Kind<G, T>`.
+
+ Specific functions are expressed as a subclass of `FunctionK`. Thus, if we have a function `f` that takes a parameter of type `FunctionK<F, G>`, we can to `f` any function from `Kind<F, T>` to `Kind<G, T>` that is polymorphic. Subclasses of `FunctionK` only need to override the generic method `invoke`, which represents the actual body of our function. `callAsFunction` is implemented automatically, so we can always call a `FunctionK` as it was a regular function.
+
+ Our `SingletonArrayFunction` defined earlier can be seen as a `FunctionK<ForId, ForArrayK>.`
+ */
+final class SingletonArrayFunction: FunctionK<ForId, ForArrayK> {
+    override func invoke<A>(_ fa: Kind<ForId, A>) -> Kind<ForArrayK, A> {
+        ArrayK([fa^.value])
+    }
+}
+/*:
+ Now we can express our `callTwice` function with regular Swift syntax:
+ */
+func callTwice<A, B>(_ f: FunctionK<ForId, ForArrayK>, _ a: A, _ b: B) -> ([A], [B]) {
+    (f(a), f(b))
```

Comment:
The result of these invocations would be an `ArrayK`, so we either change the return type of the function, or extract the array. Likewise, we need to wrap `a` and `b` in `Id` in order to invoke the function:

```suggestion
    (f(Id(a)).asArray, f(Id(b)).asArray)
```

I think this should be explained in an additional paragraph.